pandas/_libs/lib.pyx

```diff
@@ -2855,12 +2856,34 @@ no_default = _NoDefault.no_default  # Sentinel indicating the default value.
 NoDefault = Literal[_NoDefault.no_default]
 
 
+def map_infer_mask(
+        ndarray[object] arr,
+        object f,
+        const uint8_t[:] mask,
+        bint convert=True,
+        object na_value=no_default,
+        cnp.dtype dtype=np.dtype(object)
+) -> np.ndarray:
+    dummy = np.empty(0, dtype=dtype)
+    result = _map_infer_mask(
+        arr,
+        f,
+        mask,
+        dummy,
+        convert,
+        na_value,
+        dtype,
+    )
+    return result
+
+
 @cython.boundscheck(False)
 @cython.wraparound(False)
-def map_infer_mask(
+def _map_infer_mask(
         ndarray[object] arr,
         object f,
         const uint8_t[:] mask,
+        numeric_object_t[:] dummy,
```

Comment:
Does changing `cnp.dtype type=np.dtype(object)` to `numeric_object_t type = np.dtype(object)` work?

I think this is a nice change but would be nice to get rid of some of the indirection

Reply:
might be cleaner to create `result` in the calling function?

Reply:
@WillAyd: I tried this first, and it compiles but then fails when called with `TypeError: an integer is required`. In other places where we do this trick, e.g.

https://github.com/pandas-dev/pandas/blob/984d75543fce5c68f0dcc4f0b3500256e4a9c0c9/pandas/_libs/groupby.pyx#L1341-L1345

we pass a value. That's the reason I went with an array here - we can create an empty one without needing to produce a value.

@jbrockmendel - will update.